pkg/sql/importer/import_processor.go

```diff
@@ -465,14 +491,18 @@ func ingestKvs(
 			bulkSummaryMu.summary.Reset()
 			bulkSummaryMu.Unlock()
 		}
-		progCh <- prog
+		select {
+		case progCh <- prog:
+		case <-ctx.Done():
+		}
```

Comment:
we spoke about an idea where a remote processor would shut itself down if it can't push to the coordinator. That's not implemented in here, correct?

Reply:
The import processor already attempts to push a progress meta periodically. As part of that periodic push, we should see the downstream status, so I don't think we actually need an additional push loop.

Reply:
but if the processor can't communicate the with the coordinator it will not receive a downstream status, i.e. a flowCtx cancellation, right? I thought we spoke about the need for the import processor to shut itself down if it has lost connection with corrdinator.

Reply:
I'm not opposed to adding more if needed. But, I think the following is how the code stands currently:

1) Every 10s, we push an update to the progress channel.  That update will be available to the next caller of Next():

https://github.com/cockroachdb/cockroach/blob/a8b0cd9ac650505b79466ebc3b5a1564feaef29b/pkg/sql/importer/import_processor.go#L194

https://github.com/cockroachdb/cockroach/blob/a8b0cd9ac650505b79466ebc3b5a1564feaef29b/pkg/sql/importer/import_processor.go#L474-L488

2) Next() is actually called in a loop in Run():

https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/execinfra/base.go#L186-L204

3) In this loop we Push() on our destination.  I believe this will eventually make its way to the Outbox which is what sends data to the remote node.  If that Outbox encounters an error because it can't talk to the remote node, it should result in our processor seeing both a context cancellation and a call to ConsumerClosed.

Reply:
Thanks for this detailed explanation!! I took a closer look at the `DistSQLReceiver.Push()` function, and I think i've convinced myself that it will handle communication errors correctly, at least on master: 
- If the import processor attempts to send a row and the receiver's status is `NeedMoreRows`, we do seem to handle communication errors, and update the status to draing/consumer closed.

https://github.com/yuzefovich/cockroach/blob/0c1095e31cf93ea7f177f8bf1750ebab188e02d7/pkg/sql/distsql_running.go#L1335-1339

- when we push metadata however, i don't think communication errors are handled. Maybe this isn't a problem, as we're only sending metadata after the receiver's status is no longer `NeedMoreRows`.  

https://github.com/yuzefovich/cockroach/blob/0c1095e31cf93ea7f177f8bf1750ebab188e02d7/pkg/sql/distsql_running.go#L1176

I'd be curious to chat with yahor if he thinks previous versions of crdb handled comm errors like master does.

Reply:
Also, you mentioned this implementation only reduces the chances of a race. Do you have an understanding of what sorts of situations can slip past these guardrails?

Reply:
spoke offline. Here's the scenario we're worried about: 

t0: remote sends a progress update, with connection
t1: remote silently looses connection with coordinator
t2: remote sends an addsstable request, without realizing it has gone rogue
t3: coordinator sends cancel request, waits for X seconds but gets crickets from remote, and cancels the flowctx
t4: remote finally receives add sstable response, but damage is done, b/c on fail or cancel has begun

We're unsure if a remote addstable request timeout could solve this problem.